File: mac/Marquee/API/Models/AuthModels.swift
import Foundation

// Discovery, sign-in, /me and /badges. The core contract types (`User`,
// `ServerInfo`, `AuthTokenResponse`, `LoginRequest`, `SetupRequest`) predate
// this layer and live in Connection/; they're aliased here so view code can
// spell everything `API.…`.

extension API {
    /// Only a plex.tv page over https: the app hands Plex sign-in URLs to
    /// the browser, and a server (or something pretending to be one)
    /// mustn't be able to make it open a file or another app's URL scheme.
    static func plexWebURL(_ string: String) -> URL? {
        guard let url = URL(string: string), url.scheme == "https",
              let host = url.host?.lowercased(), host == "plex.tv" || host.hasSuffix(".plex.tv")
        else { return nil }
        return url
    }

    typealias User = Marquee.User
    typealias ServerInfo = Marquee.ServerInfo
    /// `POST /auth/login` and `/auth/setup`.
    typealias AuthResponse = AuthTokenResponse
    /// `{"ok": true}`
    typealias OK = OKResponse

    /// `GET /me`: the `User` plus the account's own settings.
    struct Me: Codable, Hashable, Sendable {
        let id: UUID
        let username: String
        let displayName: String?
        let role: UserRole
        let libraryOwnerId: UUID
        /// The profile photo (see `User.avatarUrl`).
        var avatarUrl: String? = nil
        let autoApproveMovies: Bool
        let autoApproveTv: Bool
        let createdAt: Date
        /// See `User.linked`; nil from older servers.
        var linked: LinkedAccounts? = nil
        /// See `User.hasPassword`; nil from older servers.
        var hasPassword: Bool? = nil
        /// 0.39+: the account's request limits; nil from older servers.
        var requestLimits: RequestLimits? = nil

        var isAdmin: Bool { role == .admin }
        var canReviewRequests: Bool { role.canReviewRequests }
        /// What the website prints: the display name, else the username.
        var label: String { displayName.nonBlank ?? username }

        var user: User {
            User(
                id: id, username: username, displayName: displayName, role: role,
                libraryOwnerId: libraryOwnerId, avatarUrl: avatarUrl,
                linked: linked, hasPassword: hasPassword
            )
        }
    }

    /// `requestLimits` on `/me` (0.39+): each type is nil when it isn't
    /// limited (always for the admin and trusted members).
    struct RequestLimits: Codable, Hashable, Sendable {
        var movie: RequestLimit?
        var tv: RequestLimit?

        init(movie: RequestLimit? = nil, tv: RequestLimit? = nil) {
            self.movie = movie
            self.tv = tv
        }

        /// The line above a member's requests (app/requests/page.tsx):
        /// "Movies: 3 of 5 requests left (every 7 days) · TV: none left
        /// until Oct 3"; nil when neither type is limited.
        func summaryLine(calendar: Calendar = .current) -> String? {
            let parts = [
                movie.map { $0.line(label: "Movies", calendar: calendar) },
                tv.map { $0.line(label: "TV", calendar: calendar) },
            ].compactMap { $0 }
            return parts.isEmpty ? nil : parts.joined(separator: " · ")
        }
    }

    /// One type's limit: at most `limit` requests in any `days` days.
    struct RequestLimit: Codable, Hashable, Sendable {
        let limit: Int
        let days: Int
        let used: Int
        let remaining: Int
        /// While none is left: when the oldest counted request ages out.
        var nextSlotAt: Date? = nil

        /// The website's `quotaLine`.
        func line(label: String, calendar: Calendar = .current) -> String {
            if remaining > 0 {
                return "\(label): \(remaining) of \(limit) requests left (every \(days) days)"
            }
            guard let nextSlotAt else { return "\(label): none left" }
            // "Oct 3", like the website's toLocaleDateString("en-US", …).
            let style = Date.FormatStyle(locale: Locale(identifier: "en_US"), calendar: calendar, timeZone: calendar.timeZone)
                .month(.abbreviated).day()
            return "\(label): none left until \(nextSlotAt.formatted(style))"
        }
    }

    // MARK: Plex / Jellyfin sign-in

    /// The media servers a Marquee account can sign in with; also the path
    /// segment in `/me/links/{server}` and `/users/import/{server}`.
    enum MediaServer: String, Codable, CaseIterable, Hashable, Sendable, Identifiable {
        case plex
        case jellyfin

        var id: String { rawValue }
        var label: String {
            switch self {
            case .plex: "Plex"
            case .jellyfin: "Jellyfin"
            }
        }
    }

    /// `POST /auth/plex/start` and `/me/links/plex/start`: open `authUrl` in
    /// the browser, then poll with `handle` (never a bare pin id) until
    /// `expiresAt`.
    struct PlexSignInStart: Codable, Hashable, Sendable {
        let handle: String
        let authUrl: String
        let expiresAt: Date

        var url: URL? { API.plexWebURL(authUrl) }
    }

    /// `POST /me/links/plex/poll` body.
    struct PlexLinkPollRequest: Codable, Hashable, Sendable {
        let handle: String
    }

    /// `POST /me/links/jellyfin` body.
    struct JellyfinLinkRequest: Codable, Hashable, Sendable {
        let username: String
        let password: String
    }

    /// `GET /me/plex-watchlist`: "Request from my Plex Watchlist", which
    /// requests what this account adds to its own Plex Watchlist.
    struct PlexWatchlist: Codable, Hashable, Sendable {
        /// Plex is linked to this account, so it can be turned on.
        let available: Bool
        let enabled: Bool
        /// Which kinds get requested.
        let movies: Bool
        let tv: Bool
        /// The last successful check; nil before the first.
        let lastSyncedAt: Date?
        /// Why the last check failed, or why it switched itself off.
        let lastError: String?
        let requestedCount: Int

        /// What an older server, without the endpoint (404), amounts to.
        static let unavailable = PlexWatchlist(
            available: false, enabled: false, movies: true, tv: true,
            lastSyncedAt: nil, lastError: nil, requestedCount: 0
        )

        /// plex-watchlist.tsx's status line: "Checked 5m ago · 3 titles
        /// requested so far", or "Checking your watchlist…" before the first.
        func statusLine(now: Date = Date()) -> String {
            var line = lastSyncedAt.map { "Checked \(Format.timeAgo($0, now: now))" } ?? "Checking your watchlist…"
            if requestedCount > 0 {
                line += " · \(requestedCount) \(requestedCount == 1 ? "title" : "titles") requested so far"
            }
            return line
        }
    }

    /// `PATCH /me/plex-watchlist` body: only the kinds being changed.
    struct PlexWatchlistTypes: Codable, Hashable, Sendable {
        var movies: Bool?
        var tv: Bool?
    }

    /// A Plex user id (a number) or a Jellyfin one (a string), sent back to
    /// the server exactly as it came.
    enum ExternalID: Codable, Hashable, Sendable, CustomStringConvertible {
        case string(String)
        case number(Int64)

        init(from decoder: Decoder) throws {
            let container = try decoder.singleValueContainer()
            if let number = try? container.decode(Int64.self) {
                self = .number(number)
            } else {
                self = .string(try container.decode(String.self))
            }
        }

        func encode(to encoder: Encoder) throws {
            var container = encoder.singleValueContainer()
            switch self {
            case let .string(value): try container.encode(value)
            case let .number(value): try container.encode(value)
            }
        }

        var description: String {
            switch self {
            case let .string(value): value
            case let .number(value): String(value)
            }
        }
    }

    /// A row in "Import from Plex/Jellyfin" (`GET /users/import/{server}`).
    struct ImportCandidate: Codable, Hashable, Sendable, Identifiable {
        let id: ExternalID
        let username: String
        let displayName: String?
        /// A picture URL on Plex/Jellyfin, when there is one.
        let thumb: String?
        /// Already linked to a Marquee account: shown, not selectable.
        let alreadyMember: Bool

        var label: String { displayName.nonBlank ?? username }

        init(from decoder: Decoder) throws {
            let container = try decoder.container(keyedBy: CodingKeys.self)
            id = try container.decode(ExternalID.self, forKey: .id)
            username = try container.decode(String.self, forKey: .username)
            displayName = try container.decodeIfPresent(String.self, forKey: .displayName)
            thumb = try container.decodeIfPresent(String.self, forKey: .thumb)
            alreadyMember = try container.decodeIfPresent(Bool.self, forKey: .alreadyMember) ?? false
        }
    }

    /// `POST /users/import/{server}` body.
    struct ImportUsersRequest: Codable, Hashable, Sendable {
        let ids: [ExternalID]
    }

    /// `POST /users/import/{server}` response.
    struct ImportUsersResult: Codable, Hashable, Sendable {
        let created: [HouseholdMember]
        let skipped: Int
    }

    /// `GET`/`PUT /settings/sign-in` (admin).
    struct SignInSettings: Codable, Hashable, Sendable {
        /// "New accounts from Plex/Jellyfin sign-in".
        var mediaServerSignup: Bool
    }

    /// `GET /badges`: the header counters, suited to polling.
    struct Badges: Codable, Hashable, Sendable {
        let unreadNotifications: Int
        /// Always 0 for members.
        let pendingRequests: Int
        /// Open problem reports (0.38+, admin; always 0 for members). nil
        /// from an older server.
        var openIssues: Int? = nil

        static let zero = Badges(unreadNotifications: 0, pendingRequests: 0)

        /// The Requests rail badge: requests and problem reports both wait
        /// on the Requests page (components/sidebar.tsx's sum).
        var requestsPageCount: Int { pendingRequests + (openIssues ?? 0) }

        /// The bell's cap on the website: "9+".
        var bellLabel: String? {
            unreadNotifications > 0 ? (unreadNotifications > 9 ? "9+" : "\(unreadNotifications)") : nil
        }

        /// The Dock tile's label: "99+".
        var dockLabel: String? {
            unreadNotifications > 0 ? (unreadNotifications > 99 ? "99+" : "\(unreadNotifications)") : nil
        }
    }

    /// `{"count": 3}` from `/requests/pending-count` and `/notifications/unread-count`.
    struct Count: Codable, Hashable, Sendable {
        let count: Int
    }

    /// `{"results": [...]}`: the lists the website shows whole (deviation 6).
    struct ListResponse<Item: Codable & Sendable>: Codable, Sendable {
        let results: [Item]
    }

    /// A movie or TV title's identity: `{"mediaType", "tmdbId"}` on the wire
    /// (`POST /surprise`, a franchise's `addAllMissing`).
    struct TitleID: Codable, Hashable, Sendable, CustomStringConvertible {
        let mediaType: MediaType
        let tmdbId: Int

        init(_ mediaType: MediaType, _ tmdbId: Int) {
            self.mediaType = mediaType
            self.tmdbId = tmdbId
        }

        /// `"movie:603"`, the website's status-map key.
        var description: String { "\(mediaType.rawValue):\(tmdbId)" }

        /// `marquee://title/movie/603`, what AppModel.handle(url:) and
        /// notification banners route with.
        var route: URL {
            URL(string: "marquee://title/\(mediaType.rawValue)/\(tmdbId)")!
        }
    }
}

extension User {
    /// What the website prints: the display name, else the username.
    var label: String { displayName.nonBlank ?? username }
}

extension String {
    /// nil when empty or whitespace-only.
    var nonBlank: String? {
        trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? nil : self
    }
}

extension Optional where Wrapped == String {
    /// nil for nil, empty or whitespace-only.
    var nonBlank: String? {
        self?.nonBlank
    }
}

File: windows/Marquee.Core/Models/AuthModels.cs
namespace Marquee.Core.Models;

// Discovery, sign-in, /me and /badges (api-v1.md section 1), plus the shared
// shapes every area uses. Keys are the server's camelCase names in PascalCase;
// Json.Options does the mapping, so no attributes are needed.
//
// Response DTOs mark every non-nullable wire field `required`: a missing key
// then fails decoding the way Swift's synthesized Decodable does, instead of
// silently reading as 0 or null. Nullable fields are plain `T?` so a key an
// older server omits reads as null.

/// <summary><c>User</c> from the contract: the signed-in account as <c>/me</c> and login return it.</summary>
public sealed record User
{
    public required Guid Id { get; init; }
    public required string Username { get; init; }
    public string? DisplayName { get; init; }

    /// <summary>Open: a role a newer server adds decodes as an unknown value instead of failing sign-in.</summary>
    public required UserRole Role { get; init; }

    /// <summary>Whose integrations and library this user sees.</summary>
    public required Guid LibraryOwnerId { get; init; }

    /// <summary>
    /// The profile photo as a server-relative path
    /// (<c>/api/v1/users/{id}/avatar?v=…</c>), fetched with the bearer token
    /// (<c>ApiClient.GetBytesAsync</c>); null when there is none, and on a
    /// server older than 0.29.0. It changes whenever the photo does, so the
    /// image can be cached under it for good.
    /// </summary>
    public string? AvatarUrl { get; init; }

    /// <summary>
    /// Which Plex/Jellyfin accounts sign in to this one; null from servers
    /// that predate Plex/Jellyfin sign-in.
    /// </summary>
    public LinkedAccounts? Linked { get; init; }

    /// <summary>
    /// False for an account made by Plex/Jellyfin sign-in or import that
    /// hasn't set a password; null from older servers (which always have one).
    /// </summary>
    public bool? HasPassword { get; init; }

    public bool IsAdmin => Role == UserRole.Admin;

    /// <inheritdoc cref="UserRole.ReviewsRequests"/>
    public bool ReviewsRequests => Role.ReviewsRequests;

    /// <summary>What the website prints: the display name, else the username.</summary>
    public string Label => DisplayName.NonBlank() ?? Username;
}

/// <summary><c>GET /me</c>: the <see cref="User"/> plus the account's own settings.</summary>
public sealed record Me
{
    public required Guid Id { get; init; }
    public required string Username { get; init; }
    public string? DisplayName { get; init; }
    public required UserRole Role { get; init; }
    public required Guid LibraryOwnerId { get; init; }

    /// <inheritdoc cref="User.AvatarUrl"/>
    public string? AvatarUrl { get; init; }

    public required bool AutoApproveMovies { get; init; }
    public required bool AutoApproveTv { get; init; }
    public required DateTimeOffset CreatedAt { get; init; }

    /// <inheritdoc cref="User.Linked"/>
    public LinkedAccounts? Linked { get; init; }

    /// <inheritdoc cref="User.HasPassword"/>
    public bool? HasPassword { get; init; }

    /// <summary>
    /// 0.39+: the account's request limits (each null when that type isn't
    /// limited, always so for the admin and trusted members); null from an
    /// older server, which has none.
    /// </summary>
    public RequestLimits? RequestLimits { get; init; }

    public bool IsAdmin => Role == UserRole.Admin;

    /// <inheritdoc cref="UserRole.ReviewsRequests"/>
    public bool ReviewsRequests => Role.ReviewsRequests;

    /// <summary>What the website prints: the display name, else the username.</summary>
    public string Label => DisplayName.NonBlank() ?? Username;

    public User User => new()
    {
        Id = Id,
        Username = Username,
        DisplayName = DisplayName,
        Role = Role,
        LibraryOwnerId = LibraryOwnerId,
        AvatarUrl = AvatarUrl,
        Linked = Linked,
        HasPassword = HasPassword,
    };
}

/// <summary><c>GET /server-info</c>, the public discovery endpoint.</summary>
public sealed record ServerInfo
{
    public const int SupportedApiVersion = 1;

    /// <summary>The first server release with the v1 API, what "Update required" asks for.</summary>
    public const string MinimumServerVersion = "0.22.0";

    /// <summary>Always <c>"marquee"</c>; anything else isn't one of ours.</summary>
    public required string App { get; init; }

    public required int ApiVersion { get; init; }

    /// <summary>The server's package.json version, e.g. <c>"0.22.0"</c>.</summary>
    public string Version { get; init; } = "unknown";

    /// <summary>Null when the server can't reach its database (<c>Status == "degraded"</c>).</summary>
    public bool? SetupComplete { get; init; }

    public string Status { get; init; } = "ok";

    /// <summary>
    /// Which sign-in methods the server offers; null from servers that
    /// predate Plex/Jellyfin sign-in, which offer only the password.
    /// </summary>
    public SignInMethods? SignIn { get; init; }

    /// <summary>"Sign in with Plex" is offered.</summary>
    public bool OffersPlexSignIn => SignIn?.Plex == true;

    /// <summary>"Sign in with Jellyfin" is offered.</summary>
    public bool OffersJellyfinSignIn => SignIn?.Jellyfin == true;

    /// <summary>
    /// What to call the "jellyfin" server in the UI: "Emby" when that's what
    /// is connected (it speaks the same API), otherwise "Jellyfin" — also for
    /// servers before 0.40, which don't send it.
    /// </summary>
    public string JellyfinName => SignIn?.JellyfinName ?? MediaServerKindExtensions.DefaultJellyfinName;

    /// <summary>The user-facing name of <paramref name="server"/> on this server.</summary>
    public string MediaServerName(MediaServerKind server) => server.Label(JellyfinName);

    /// <summary>
    /// The line under the Plex/Jellyfin buttons telling a newcomer how to
    /// get an account, when the admin has new accounts from sign-in on;
    /// names only the methods that make one, single sign-on first as on the
    /// website ("Authentik (or Plex or Jellyfin)"). Null otherwise (and from
    /// servers older than 0.43, which don't say).
    /// </summary>
    public string? SignupHint
    {
        get
        {
            if (SignIn is not { } signIn) return null;
            var names = new List<string>();
            if (signIn.Sso is { Signup: true } sso) names.Add(sso.Name);
            if (signIn.Signup && signIn.Plex) names.Add("Plex");
            if (signIn.Signup && signIn.Jellyfin) names.Add(JellyfinName);
            if (names.Count == 0) return null;
            var named = names.Count == 1 ? names[0] : $"{names[0]} (or {string.Join(" or ", names.Skip(1))})";
            return $"New here? Use Sign in with {named} — your account is made for you.";
        }
    }

    /// <summary>"Sign in with {name}" for the admin's single sign-on is offered (0.44+).</summary>
    public bool OffersSsoSignIn => SignIn?.Sso != null;

    /// <summary>The single sign-on button's name ("Authentik"); null while it isn't offered.</summary>
    public string? SsoName => SignIn?.Sso?.Name;

    /// <summary>"Use Quick Connect" is offered on the Jellyfin sign-in (0.44+; never for Emby).</summary>
    public bool OffersQuickConnect => SignIn?.QuickConnect == true && OffersJellyfinSignIn;

    public bool IsMarquee => App == "marquee";
    public bool IsSupported => ApiVersion == SupportedApiVersion;
    public bool IsDegraded => Status == "degraded";
}

/// <summary><c>POST /auth/login</c> and <c>/auth/setup</c> response.</summary>
public sealed record AuthResponse
{
    /// <summary><c>mqt_</c> + 43 base64url characters.</summary>
    public required string Token { get; init; }

    public required DateTimeOffset ExpiresAt { get; init; }
    public required User User { get; init; }
}

public sealed record LoginRequest(string Username, string Password, string DeviceName);

/// <summary>
/// <c>server-info.signIn</c>: the sign-in methods this server offers. Plex
/// and Jellyfin appear only while that integration is connected.
/// </summary>
public sealed record SignInMethods
{
    public bool Password { get; init; } = true;
    public bool Plex { get; init; }
    public bool Jellyfin { get; init; }

    private readonly string jellyfinName = MediaServerKindExtensions.DefaultJellyfinName;

    /// <summary>
    /// <c>jellyfinName</c> (0.40+): "Jellyfin" or "Emby". Missing or blank
    /// (older servers) reads as "Jellyfin".
    /// </summary>
    public string JellyfinName
    {
        get => jellyfinName;
        init => jellyfinName = MediaServerKindExtensions.NormalizedJellyfinName(value);
    }

    /// <summary>
    /// <c>signup</c> (0.42.2+): new accounts from Plex/Jellyfin sign-in are on
    /// (and one of them is offered). Missing (older servers) reads as false.
    /// </summary>
    public bool Signup { get; init; }

    /// <summary>
    /// <c>quickConnect</c> (0.44+): Jellyfin sign-in is on and the server is
    /// Jellyfin, not Emby. Missing (older servers) reads as false.
    /// </summary>
    public bool QuickConnect { get; init; }

    /// <summary>
    /// <c>sso</c> (0.44+): single sign-on is set up. Null when it isn't, and
    /// from older servers, which don't send it.
    /// </summary>
    public SsoSignIn? Sso { get; init; }
}

public sealed record SetupRequest(string Username, string Password, string DisplayName, string DeviceName);

/// <summary><c>{"ok": true}</c></summary>
public sealed record OK
{
    public required bool Ok { get; init; }
}

/// <summary>For calls whose body the caller doesn't need (or a 204). An empty body decodes to it without parsing.</summary>
public readonly record struct EmptyResponse;

/// <summary><c>{"count": 3}</c> from <c>/requests/pending-count</c> and <c>/notifications/unread-count</c>.</summary>
public sealed record CountResponse
{
    public required int Count { get; init; }
}

// Both list wrappers constrain T to notnull: an entry is never null on the
// wire, and the constraint is what lets Json's null check read that off the
// element type (an unconstrained T is annotated as maybe-null in metadata).

/// <summary><c>{"page", "totalPages", "totalResults", "results"}</c>: the Movies/Series grids.</summary>
public sealed record Paginated<T> where T : notnull
{
    public required int Page { get; init; }
    public required int TotalPages { get; init; }
    public required int TotalResults { get; init; }
    public required IReadOnlyList<T> Results { get; init; }

    public bool HasMorePages => Page < TotalPages;
}

/// <summary><c>{"results": [...]}</c>: the lists the website shows whole (deviation 6).</summary>
public sealed record ListResponse<T> where T : notnull
{
    public required IReadOnlyList<T> Results { get; init; }
}

/// <summary><c>GET /badges</c>: the header counters, suited to polling.</summary>
public sealed record Badges
{
    public required int UnreadNotifications { get; init; }

    /// <summary>Always 0 for members.</summary>
    public required int PendingRequests { get; init; }

    /// <summary>Open problem reports (0.38+; an older server omits it, meaning 0). Always 0 for members.</summary>
    public int OpenIssues { get; init; }

    /// <summary>The Requests badge: pending requests plus open problem reports, since both wait on that page.</summary>
    public int RequestsBadge => PendingRequests + OpenIssues;

    public static Badges Zero { get; } = new() { UnreadNotifications = 0, PendingRequests = 0 };

    /// <summary>The bell's cap on the website: "9+".</summary>
    public string? BellLabel =>
        UnreadNotifications > 0 ? (UnreadNotifications > 9 ? "9+" : UnreadNotifications.ToString()) : null;

    /// <summary>The taskbar badge's cap: "99+".</summary>
    public string? TaskbarLabel =>
        UnreadNotifications > 0 ? (UnreadNotifications > 99 ? "99+" : UnreadNotifications.ToString()) : null;
}

/// <summary>
/// A movie or TV title's identity: <c>{"mediaType", "tmdbId"}</c> on the wire
/// (<c>POST /surprise</c>, a franchise's <c>addAllMissing</c>).
/// </summary>
public readonly record struct TitleId(MediaType MediaType, int TmdbId)
{
    /// <summary><c>"movie:603"</c>, the website's status-map key.</summary>
    public override string ToString() => $"{MediaType.Value}:{TmdbId}";

    /// <summary><c>marquee://title/movie/603</c>, what notification banners and protocol activation route with.</summary>
    public Uri Route => new($"marquee://title/{MediaType.Value}/{TmdbId}");
}
